src/lib.rs

```diff
@@ -0,0 +1,350 @@
+extern crate core;
+
+pub mod commands;
+
+use crate::commands::CargoCommand;
+use cargo_metadata::{Message, MetadataCommand};
+use core::fmt;
+use rustc_version::Channel;
+use semver::Version;
+use serde::Deserialize;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus, Stdio};
+use std::{env, io, process};
+use tee::TeeReader;
+
+pub fn build_elf(
+    cmd: CargoCommand,
+    message_format: &str,
+    args: &Vec<String>,
+) -> (ExitStatus, Vec<Message>) {
+    let mut command = make_cargo_build_command(cmd, message_format, args);
+    let mut process = command.spawn().unwrap();
+    let command_stdout = process.stdout.take().unwrap();
+
+    let mut tee_reader;
+    let mut stdout_reader;
+    let buf_reader: &mut dyn BufRead = if message_format == "json-render-diagnostics" {
+        stdout_reader = BufReader::new(command_stdout);
+        &mut stdout_reader
+    } else {
+        tee_reader = BufReader::new(TeeReader::new(command_stdout, io::stdout()));
+        &mut tee_reader
+    };
+
+    let messages = Message::parse_stream(buf_reader)
+        .collect::<io::Result<_>>()
+        .unwrap();
+
+    (process.wait().unwrap(), messages)
+}
+
+fn make_cargo_build_command(
+    cmd: CargoCommand,
+    message_format: &str,
+    args: &Vec<String>,
+) -> Command {
+    let rust_flags = env::var("RUSTFLAGS").unwrap_or_default()
+        + &format!(
+            " -L{}/libctru/lib -lctru",
+            env::var("DEVKITPRO").expect("DEVKITPRO is not defined as an environment variable")
+        );
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let sysroot = find_sysroot();
+    let mut command = Command::new(cargo);
+
+    if !sysroot.join("lib/rustlib/armv6k-nintendo-3ds").exists() {
+        eprintln!("No pre-build std found, using build-std");
+        command.arg("-Z").arg("build-std");
+    }
+
+    println!("{:?}", args);
```

Comment:
Is this just for debugging? Let's remove it before we merge.

Reply:
Oops that's my bad. Accidentally left over from testing. Will remove 👍